src/test/java/com/fasterxml/jackson/databind/jsontype/TestPolymorphicDeserialization5016.java

```diff
@@ -0,0 +1,96 @@
+package com.fasterxml.jackson.databind.jsontype;
+
+import com.fasterxml.jackson.annotation.JsonTypeInfo;
+import com.fasterxml.jackson.databind.MapperFeature;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.databind.exc.InvalidDefinitionException;
+import com.fasterxml.jackson.databind.exc.InvalidTypeIdException;
+import com.fasterxml.jackson.databind.json.JsonMapper;
+import com.fasterxml.jackson.databind.testutil.DatabindTestUtil;
+import org.junit.jupiter.api.Test;
+
+import java.io.IOException;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+
+// https://github.com/FasterXML/jackson-databind/issues/5016
+public class TestPolymorphicDeserialization5016 extends DatabindTestUtil
+{
+    static abstract class Animal {
+        public String name = "animal";
+    }
+
+    static abstract class Plant {
+        public String name = "plant";
+    }
+
+    static class Cat extends Animal {
+        public String name = "cat";
+    }
+
+    static class Dog extends Animal implements Runnable {
+        public String name = "dog";
+
+        @Override
+        public void run() {
+            System.out.println("Dog is running");
+        }
+    }
+
+    static class Tree extends Plant {
+        public String name = "tree";
+    }
+
+    static class AnimalInfo {
+        @JsonTypeInfo(use = JsonTypeInfo.Id.CLASS,
+                include = JsonTypeInfo.As.PROPERTY,
+                property = "@class")
+        public Animal thisType;
+    }
+
+    static class PlantInfo {
+        @JsonTypeInfo(use = JsonTypeInfo.Id.CLASS,
+                include = JsonTypeInfo.As.PROPERTY,
+                property = "@class")
+        public Plant thisType;
+    }
+
+    static class RunnableInfo {
+        @JsonTypeInfo(use = JsonTypeInfo.Id.CLASS,
+                include = JsonTypeInfo.As.PROPERTY,
+                property = "@class")
+        public Runnable thisType;
+    }
+
+    @Test
+    public void testWrongSubtype() throws IOException {
```

Comment:
Oh. I was thinking of somewhat simpler set up: actually simply replacing class name with something unrelated.
But this works too.